File: business/src/main/java/com/radioteria/business/services/user/impl/UserServiceImpl.java
package com.radioteria.business.services.user.impl;

import com.radioteria.business.events.userService.PasswordChangedEvent;
import com.radioteria.business.events.userService.UserConfirmedEvent;
import com.radioteria.business.events.userService.UserRegisteredEvent;
import com.radioteria.business.services.user.api.UserService;
import com.radioteria.business.services.user.exceptions.UserExistsException;
import com.radioteria.business.services.user.exceptions.UserNotFoundException;
import com.radioteria.db.dao.api.UserDao;
import com.radioteria.db.entities.User;
import com.radioteria.db.enumerations.UserState;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationEvent;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Optional;


@Service
public class UserServiceImpl implements UserService {

    private final UserDao userDao;

    private final PasswordEncoder passwordEncoder;

    private final ApplicationEventPublisher eventPublisher;

    @Value("${registration.email.verify.enabled}")
    private boolean emailVerifyEnabled = true;

    @Autowired
    public UserServiceImpl(UserDao userDao,
                           PasswordEncoder passwordEncoder,
                           ApplicationEventPublisher eventPublisher) {

        this.userDao = userDao;
        this.passwordEncoder = passwordEncoder;
        this.eventPublisher = eventPublisher;

    }

    @Override
    public Optional<User> findByEmail(String email) {

        return userDao.findByEmail(email);

    }

    @Override
    public Optional<User> findById(Long id) {

        return userDao.find(id);

    }

    public boolean passwordMatches(User user, String plainPassword) {

        return passwordEncoder.matches(plainPassword, user.getPassword());

    }

    public boolean isEmailAvailable(String email) {

        return userDao.isEmailAvailable(email);

    }

    public void register(String email, String plainPassword, String name) {

        throwErrorIfEmailNotAvailable(email);

        String encodedPassword = passwordEncoder.encode(plainPassword);

        User user = new User();

        user.setEmail(email);
        user.setPassword(encodedPassword);
        user.setName(name);
        user.setState(emailVerifyEnabled ? UserState.INACTIVE : UserState.ACTIVE);
        user.setAvatarFile(null);

        userDao.persist(user);

        publishEventAboutUserRegistered(user);

    }

    public void activateByEmail(String email) {

        User user = findByEmail(email).orElseThrow(() ->
                new UserNotFoundException(String.format("User with email \"%s\" not exists.", email)));

        user.setState(UserState.ACTIVE);

        publishEventAboutUserConfirmed(user);

    }

    public void changePassword(User user, String newPlainPassword) {

        String encodedPassword = passwordEncoder.encode(newPlainPassword);

        user.setPassword(encodedPassword);

        publishEventAboutChangedPassword(user);

    }

    public void deactivate(User user) {

        user.setState(UserState.DELETED);

        publishEventAboutUserDeleted(user);

    }

    public void delete(User user) {

        throw new IllegalStateException("No, God! Please, No!");

    }

    private void throwErrorIfEmailNotAvailable(String email) {

        if (!isEmailAvailable(email)) {
            throw new UserExistsException(String.format("User with email \"%s\" already exists.", email));
        }

    }

    private void publishEventAboutUserRegistered(User user) {

        ApplicationEvent event = new UserRegisteredEvent(this, user);

        eventPublisher.publishEvent(event);

    }

    private void publishEventAboutUserConfirmed(User user) {

        ApplicationEvent event = new UserConfirmedEvent(this, user);

        eventPublisher.publishEvent(event);

    }

    private void publishEventAboutUserDeleted(User user) {

        ApplicationEvent event = new UserConfirmedEvent(this, user);

        eventPublisher.publishEvent(event);

    }

    private void publishEventAboutChangedPassword(User user) {

        ApplicationEvent event = new PasswordChangedEvent(this, user);

        eventPublisher.publishEvent(event);

    }
}

File: business/src/main/kotlin/com/radioteria/services/user/impl/UserServiceImpl.kt
package com.radioteria.services.user.impl

import com.radioteria.db.repositories.UserRepository
import com.radioteria.services.user.events.PasswordChangedEvent
import com.radioteria.services.user.events.UserConfirmedEvent
import com.radioteria.services.user.events.UserRegisteredEvent
import com.radioteria.services.user.UserService
import com.radioteria.db.entities.User
import com.radioteria.db.enums.UserState
import com.radioteria.services.user.events.UserDeletedEvent
import com.radioteria.services.user.exceptions.UserServiceException
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.ApplicationEventPublisher
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserServiceImpl(private val userRepository: UserRepository,
                      private val passwordEncoder: PasswordEncoder,
                      private val eventPublisher: ApplicationEventPublisher
) : UserService {

    @Value("\${registration.email.verify.enabled}")
    var emailVerifyEnabled = true

    override fun findByEmail(email: String): User? {
        return userRepository.findByEmail(email)
    }

    override fun findById(id: Long): User? {
        return userRepository.findById(id)
    }

    override fun isPasswordMatch(plainPassword: String, user: User): Boolean {
        return passwordEncoder.matches(plainPassword, user.password)
    }

    override fun isEmailAvailable(email: String): Boolean {
        return userRepository.isEmailAvailable(email)
    }

    override fun register(email: String, plainPassword: String, name: String) {
        throwErrorIfEmailNotAvailable(email)

        val encodedPassword = passwordEncoder.encode(plainPassword)
        val user = User(email = email, password = encodedPassword, name = name,
                state = if (emailVerifyEnabled) UserState.INACTIVE else UserState.ACTIVE)

        userRepository.persist(user)

        eventPublisher.publishEvent(UserRegisteredEvent(this, user))
    }

    override fun activateByEmail(email: String) {
        val user = findByEmail(email)
                ?: throw UserServiceException(String.format("User with email \"%s\" not exists.", email))

        user.state = UserState.ACTIVE

        eventPublisher.publishEvent(UserConfirmedEvent(this, user))
    }

    override fun changePassword(user: User, newPlainPassword: String) {
        val encodedPassword = passwordEncoder.encode(newPlainPassword)

        user.password = encodedPassword

        eventPublisher.publishEvent(PasswordChangedEvent(this, user))
    }

    override fun deactivate(user: User) {
        user.state = UserState.DELETED

        eventPublisher.publishEvent(UserDeletedEvent(this, user))
    }

    override fun delete(user: User) {
        throw IllegalStateException("No, God! Please, No!")
    }

    private fun throwErrorIfEmailNotAvailable(email: String) {
        if (!isEmailAvailable(email)) {
            throw UserServiceException(String.format("User with email \"%s\" already exists.", email))
        }
    }

}
